src/sinks/util/udp.rs

```diff
@@ -83,103 +92,96 @@ impl UdpSink {
+                                        self.next_delay().await
+                                    }
+                                }
+                            }
+                            None => {
+                                error!(message = "DNS resolved no addresses", host = %self.host);
+                                self.next_delay().await
                             }
                         },
                         Err(error) => {
-                            error!(message = "unable to bind local address", addr = %bind_address, %error);
-                            State::Backoff(self.next_delay01())
+                            error!(message = "unable to resolve DNS", host = %self.host, %error);
+                            self.next_delay().await
                         }
                     }
                 }
-                State::Connected(ref mut socket) => return Ok(Async::Ready(socket)),
-                State::Backoff(ref mut delay) => match delay.poll() {
-                    Ok(Async::NotReady) => return Ok(Async::NotReady),
-                    Ok(Async::Ready(())) => State::Initializing,
-                    Err(_) => unreachable!(),
-                },
             }
         }
     }
 }
 
-fn find_bind_address(remote_addr: &SocketAddr) -> SocketAddr {
-    match remote_addr {
-        SocketAddr::V4(_) => SocketAddr::new(IpAddr::V4(Ipv4Addr::UNSPECIFIED), 0),
-        SocketAddr::V6(_) => SocketAddr::new(IpAddr::V6(Ipv6Addr::UNSPECIFIED), 0),
-    }
-}
-
-impl Sink for UdpSink {
-    type SinkItem = Bytes;
-    type SinkError = ();
+#[async_trait]
+impl StreamSink for UdpSink {
+    async fn run(&mut self, mut input: BoxStream<'_, Event>) -> Result<(), ()> {
+        while let Some(event) = input.next().await {
+            let mut bytes = match encode_event(event, &self.encoding) {
+                Some(bytes) => bytes,
+                None => continue,
+            };
 
-    fn start_send(&mut self, line: Self::SinkItem) -> StartSend<Self::SinkItem, Self::SinkError> {
-        let span = self.span.clone();
-        let _enter = span.enter();
+            let span = self.span.clone();
+            async {
+                let socket = self.get_socket().await;
 
-        match self.poll_inner() {
-            Ok(Async::Ready(socket)) => {
                 debug!(
                     message = "sending event.",
-                    bytes = &field::display(line.len())
+                    bytes = &field::display(bytes.len())
                 );
-                match socket.send(&line) {
-                    Err(error) => {
-                        self.state = State::Backoff(self.next_delay01());
-                        error!(message = "send failed", %error);
-                        Ok(AsyncSink::NotReady(line))
+                while !bytes.is_empty() {
+                    match socket.send(&bytes).await {
+                        Ok(cnt) => bytes.advance(cnt),
```

Comment:
UDP does not guarantee the order of packets, so we cannot split one message into multiple packets, I added a warning for this in #4034.